File: My_Notes/Programming_Note/Samples/async_samples/go/sync.go
package main

import (
	"fmt"
	"time"
)

func syncFunc() {
	job := job{
		db:  3,
		api: 1,
		ui:  2,
	}
	var dbEngineer dbEngineer
	var apiEngineer apiEngineer
	var uiEngineer uiEngineer

	dbEngineer.do(job)
	apiEngineer.do(job)
	uiEngineer.do(job)
}

type engineer interface {
	do(job)
}

type dbEngineer struct{}

func (dbEngineer) do(job job) {
	fmt.Println("db task start")
	time.Sleep(time.Second * time.Duration(job.db))
	fmt.Println("db task complete")
}

type apiEngineer struct{}

func (apiEngineer) do(job job) {
	fmt.Println("api task start")
	time.Sleep(time.Second * time.Duration(job.api))
	fmt.Println("api task complete")
}

type uiEngineer struct{}

func (uiEngineer) do(job job) {
	fmt.Println("ui task start")
	time.Sleep(time.Second * time.Duration(job.ui))
	fmt.Println("ui task complete")
}

File: My_Notes/Programming_Note/Samples/async_samples/ts/sync.ts
function sync(): void {
    var job = new Job(3,1,2);
    var dbEngineer = new DBEngineer();
    var apiEngineer = new APIEngineer();
    var uiEngineer = new UIEngineer();
    dbEngineer.do(job).then(
        ()=>{apiEngineer.do(job).then(
            ()=>{uiEngineer.do(job)}
        );}
    );
}

class Job {
    db:number;
    api:number;
    ui:number;
    constructor(db:number, api:number, ui:number){
        this.db = db;
        this.api = api;
        this.ui = ui;
    }
}

interface IEngineer {
    do(job: Job): void;
}

class DBEngineer implements IEngineer {
    async do(job:Job){
        console.log("db task start")
        await new Promise(f => setTimeout(f, 1000 * job.db));
        console.log("db task complete")
    }
}

class APIEngineer implements IEngineer {
    async do(job:Job){
        console.log("api task start")
        await new Promise(f => setTimeout(f, 1000 * job.api));
        console.log("api task complete")
    }
}

class UIEngineer implements IEngineer {
    async do(job:Job){
        console.log("ui task start")
        await new Promise(f => setTimeout(f, 1000 * job.ui));
        console.log("ui task complete")
    }
}